src/core/api.rs

```diff
@@ -174,99 +178,105 @@ impl Covercrypt {
-        let mut res =
-            Vec::with_capacity(plaintext.len() + Aes256Gcm::MAC_LENGTH + Aes256Gcm::NONCE_LENGTH);
-        res.extend(nonce.0);
+        let nonce = &self::Nonce::<{ Self::NONCE_LENGTH }>::new(&mut *rng);
+        let aes = Self::new(key);
+        let mut ciphertext = aes
+            .encrypt(nonce, ptx, ad)
+            .map_err(Error::CryptoCoreError)?;
+        let mut res = Vec::with_capacity(ptx.len() + Self::MAC_LENGTH + Self::NONCE_LENGTH);
+        res.extend(&Nonce::to_bytes(nonce));
         res.append(&mut ciphertext);
         Ok(res)
     }
 
-    /// Decrypts the given ciphertext using the given symmetric key.
-    ///
-    /// The encryption scheme used is AES-256 GCM.
-    ///
-    /// - `symmetric_key`   : AES key
-    /// - `ciphertext`      : encrypted data
-    /// - `ad`              : associated data
-    pub fn dem_decrypt(
-        &self,
-        symmetric_key: &SymmetricKey<SEED_LENGTH>,
-        ciphertext: &[u8],
+    fn decrypt(
+        key: &Secret<{ Self::KEY_LENGTH }>,
+        ctx: &[u8],
         ad: Option<&[u8]>,
     ) -> Result<Vec<u8>, Error> {
-        let aes256gcm = Aes256Gcm::new(symmetric_key);
-        aes256gcm
-            .decrypt(
-                &Nonce::try_from_slice(&ciphertext[..Aes256Gcm::NONCE_LENGTH])?,
-                &ciphertext[Aes256Gcm::NONCE_LENGTH..],
-                ad,
-            )
+        if ctx.len() < Self::NONCE_LENGTH {
+            return Err(Error::CryptoCoreError(
+                cosmian_crypto_core::CryptoCoreError::DecryptionError,
+            ));
+        }
+        let sym_key = SymmetricKey::try_from_slice(key)?;
+        let nonce =
+            &Nonce::<{ Self::NONCE_LENGTH }>::try_from_slice(&ctx[..{ Self::NONCE_LENGTH }])?;
+        let aes = Self::new(&sym_key);
+        aes.decrypt(nonce, &ctx[{ Self::NONCE_LENGTH }..], ad)
             .map_err(Error::CryptoCoreError)
     }
 }
 
-/// Encrypted header holding a `Covercrypt` encapsulation of a symmetric key and
-/// additional data encrypted using the `Covercrypt` DEM with the encapsulated
-/// key.
-///
-/// *Note*: the DEM ciphertext is also used to select the correct symmetric key
-/// from the decapsulation.
-///
-/// - `encapsulation`       :   `Covercrypt` encapsulation of a symmetric key
-/// - `encrypted_metadata`  :   AES-256 GCM encryption of the metadata
+/// Encrypted header holding a `Covercrypt` encapsulation of a 256-bit seed and
+/// metadata encrypted under the scheme `E` using a key derived from the
+/// encapsulated seed.
 #[derive(Debug, PartialEq, Eq)]
-pub struct EncryptedHeader {
+pub struct EncryptedHeader<
+    E: AE<KEY_LENGTH, NONCE_LENGTH, MAC_LENGTH>,
+    const KEY_LENGTH: usize,
+    const NONCE_LENGTH: usize,
+    const MAC_LENGTH: usize,
+> {
     pub encapsulation: Encapsulation,
     pub encrypted_metadata: Option<Vec<u8>>,
+    pub phantom: PhantomData<(E, usize, usize, usize)>,
```

Comment:
You only need to use generic types:
```suggestion
    pub phantom: PhantomData<E>,
```